tests/test_virus_scan_s3_bucket.py

```diff
@@ -135,87 +142,114 @@ def _get_mock_clients(self, versions_responses, versions_page_size):
     @pytest.mark.parametrize("dry_run", (False, True,))
-    def test_unfiltered(self, versions_page_size, dry_run):
+    def test_unfiltered(self, versions_page_size, dry_run, concurrency):
         av_api_client, s3_client = self._get_mock_clients(self.versions_responses, versions_page_size)
 
-        retval = virus_scan_bucket(
-            s3_client,
-            av_api_client,
-            "spade",
-            prefix="",
-            since=None,
-            dry_run=dry_run,
-        )
+        with ThreadPoolExecutor(max_workers=concurrency) if concurrency else nullcontext() as executor:
+            map_callable = map if executor is None else executor.map
+            retval = virus_scan_bucket(
+                s3_client,
+                av_api_client,
+                "spade",
+                prefix="",
+                since=None,
+                dry_run=dry_run,
+                map_callable=map_callable,
+            )
 
-        assert av_api_client.mock_calls == [] if dry_run else [
-            mock.call.scan_and_tag_s3_object("spade", "sandman/4321-billy-winks.pdf", "oo_.BepoodlLml"),
-            mock.call.scan_and_tag_s3_object("spade", "sandman/4321-billy-winks.pdf", "moB_eLplool.do"),
-            mock.call.scan_and_tag_s3_object("spade", "sandman/4321-billy-winks.pdf", "ooBmo_pe.ldoLl"),
-            mock.call.scan_and_tag_s3_object("spade", "sandman/1234-deedaw.pdf", "epmlLoBodo_ol."),
-            mock.call.scan_and_tag_s3_object("spade", "sandman/1234-deedaw.pdf", "loleLoooBp_md."),
-            mock.call.scan_and_tag_s3_object("spade", "sandman/4321-billy-winks.pdf", "molo.oB_oLdelp"),
-            mock.call.scan_and_tag_s3_object("spade", "dribbling/bib.jpeg", "ldmoo_.pBeolLo"),
-        ]
         assert s3_client.mock_calls == [
             mock.call.get_paginator("list_object_versions"),
             mock.call.get_paginator().paginate(Bucket="spade", Prefix=""),
         ]
-        assert retval == (7, 0, 0, 0,) if dry_run else (7, 4, 1, 2,)
 
+        if dry_run:
+            assert av_api_client.mock_calls == []
+            assert retval == (7, 0, 0, 0,)
+        else:
+            assert av_api_client.mock_calls == [
+                mock.call.scan_and_tag_s3_object("spade", "sandman/4321-billy-winks.pdf", "oo_.BepoodlLml"),
+                mock.call.scan_and_tag_s3_object("spade", "sandman/4321-billy-winks.pdf", "moB_eLplool.do"),
+                mock.call.scan_and_tag_s3_object("spade", "sandman/4321-billy-winks.pdf", "ooBmo_pe.ldoLl"),
+                mock.call.scan_and_tag_s3_object("spade", "sandman/1234-deedaw.pdf", "epmlLoBodo_ol."),
+                mock.call.scan_and_tag_s3_object("spade", "sandman/1234-deedaw.pdf", "loleLoooBp_md."),
+                mock.call.scan_and_tag_s3_object("spade", "sandman/4321-billy-winks.pdf", "molo.oB_oLdelp"),
+                mock.call.scan_and_tag_s3_object("spade", "dribbling/bib.jpeg", "ldmoo_.pBeolLo"),
```

Comment:
I'm curious... do the test threads run deterministically?

Reply:
Well caught - I had forgotten to put the re-sorting in. Have now. Think it was just passing out of luck.